compiler/passes/denormalize.cpp

```diff
@@ -0,0 +1,457 @@
+        }
+        denormalizeOrDeferCandidates(candidates, deferredSyms);
+
+        isFirstRound = false;
+      } while(deferredSyms.count() > 0);
+    }
+  }
+}
+
+void denormalizeOrDeferCandidates(UseDefCastMap& candidates,
+    Vec<Symbol*>& deferredSyms) {
+
+  for(UseDefCastMap::iterator it = candidates.begin() ;
+      it != candidates.end() ; ++it) {
+    // unpack the bundle
+    std::pair<Expr*, Type*> defCastPair = it->second;
+    SymExpr* use = it->first;
+    Expr* def = defCastPair.first;
+    Type* castTo = defCastPair.second;
+
+    // if parent of def is gone, it means it has been denormalized in the
+    // earlier passes
+    if(def->parentExpr == NULL) {
+      deferredSyms.add(use->var);
+      continue;
+    }
+    denormalize(def, use, castTo);
+  }
+}
+
+void findCandidatesInFunc(FnSymbol* fn, Vec<Symbol*> symVec,
+    UseDefCastMap& udcMap) {
+
+  Map<Symbol*,Vec<SymExpr*>*> defMap;
+  Map<Symbol*,Vec<SymExpr*>*> useMap;
+
+  buildDefUseMaps(fn, defMap, useMap);
+
+  bool cachedGlobalManip = isRegisteredGlobalManip(fn);
+  // Note the different symbol set being iterated
+  forv_Vec(Symbol, sym, symVec) {
+
+    SymExpr *use = NULL;
+    Expr *usePar = NULL;
+    Expr *def = NULL;
+    Expr *defPar;
+    Type* castTo = NULL;
+
+    //if we don't already have it cached,
+    //check for global symbols in function body
+    if(!cachedGlobalManip) {
+      //I am not sure if there should be !isConstant and !isParam as well
+      //I think there can be a const but stateful global that is manipulated by
+      //the function so, play safe
+      if(sym && !sym->isImmediate() && isGlobal(sym)){
+        registerGlobalManip(fn, true);
+        cachedGlobalManip = true;
+      }
+    }
+
+    if(isDenormalizable(sym, defMap, useMap, &use, &def, &castTo)) {
+      usePar = use->parentExpr;
+      defPar = def->parentExpr;
+
+      // Initially I used to defer denormalizing actuals and have special
+      // treatment while denormalizing actuals of a function call. Main reason
+      // behind that was C standard not specifying evaluation order of actuals
+      // and I wanted to keep the call order in the IR. However, if we have
+      // strong enough safety checks on `Expr`s that we are moving, than that
+      // shouldn't necessarily matter.
+      //
+      // Possible alternative for an easy-to-implement actual denormalization is
+      // to denormalize an actual only if it's the last/only one.
+
+      //denormalize if the def is safe to move and there is no unsafe function
+      //between use and def
+      if(exprHasNoSideEffects(def)) {
+        if(!unsafeExprInBetween(defPar, usePar)) {
+          std::pair<Expr*, Type*> defCastPair(def, castTo);
+          udcMap.insert(std::pair<SymExpr*, std::pair<Expr*, Type*> >
+              (use, defCastPair));
+        }
+      }
+    }
+  } // end loop for symbol
+  if(!cachedGlobalManip) {
+    registerGlobalManip(fn, false);
```

Comment:
I don't understand this bit... why would a function no longer manipulate globals if you got this far? I'd expect  the exprAnalysis.cpp to independently analyze functions (independent of denormalization) to find if they could be called in a side-effect-free way.